VRRP v1.6 support for ifupdown-ng

EasyNetDev opened the pull request:
VRRP support for ifupdown-ng.

Feature:

    Generation of subinterfaces for VRRP IPv4 and IPv6 (you need separate (yes, we need separate interface for each protocol because the MAC address is different, more details in rfc5798bis).

    IPv4 case: 00-00-5E-00-01-{VRID} (in hex, in Internet-standard bit- order)
    IPv6 case: 00-00-5E-00-02-{VRID} (in hex, in Internet-standard bit- order)

    Interface generation format is vXXXXXXXX-YY-V, where XXXXXXXX is a CRC32 checksum of the string of main interface. It can be obtained using: echo -n "eth0" | cksum or if you compiled ifupdown-ng with CRC32 support echo -n "eth0" | crc32sum. The output will be a number 1136632051 which converted in hexa with printf "%08x\n" 1136632051 we will get value 43bfa0f3.
    YY is the VRID in hexa and V is the protocol name.
    In this way we will assure that we will have the same VRRP sub-interfaces all the time after reboots and we don't have to be worry about the length of the interfaces which is limited to 15 characters.
    The executor can use two CRC32 tools: crc32sum which cames with this patch and cksum which cames with coreutils. In case cksum is not available you can build ifupdown-ng with crc32sum which is a stripped version of cksum.
    VRRP control is done in a separate daemon like FRRouting which has VRRP support.
    Minimum supported Linux Kernel is 5.1 because in this version protodown feature was introduced.

Neustradamus commented:
To follow!

Neustradamus commented:
Dear @ifupdown-ng team, @kaniini, @BarbarossaTM: What do you think?

It replaces:
- https://github.com/ifupdown-ng/ifupdown-ng/pull/222
- https://github.com/ifupdown-ng/ifupdown-ng/pull/223
- https://github.com/ifupdown-ng/ifupdown-ng/pull/232

EasyNetDev commented:
OK, I've fixed with:

```
    printf("%lu %llu\n", (long unsigned)crc, (long long unsigned)total_bytes);
```
I've tried to use format dependency of __WORDSIZE, but looks that is not working correctly on apline linux.

EasyNetDev commented:
Finally I came with a solution for compatibility compilation between 64bit and 32bit without throwing warnings:

```
    printf(PRIuFAST32 " " PRIuMAX "\n", crc, total_bytes);
```

Now we have the latest PR which is compiling without warnings.

Neustradamus commented:
Dear @kaniini, @BarbarossaTM, @awilfox: Have you looked this @EasyNetDev PR?

EasyNetDev commented:
I'm still waiting for a feedback. I already pushed those PRs in my branch and I compiled a Debian package and I'm using on my routers.

awilfox commented on tools/cksum.c:
```diff
@@ -0,0 +1,129 @@
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
```
I don't believe it would be possible to merge the code from coreutils because the GPL-3 code is not compatible with the license of ifupdown-ng itself.

EasyNetDev replied:
Hi,

And what is the solution for this? I took parts of the cksum CRC32 tool and adapted here because on embedded devices where you need to strip a lot of thinks from the OS there is a possibility for not having cksum tool. CRC32 implementation in `sh` is almost impossible.

EasyNetDev replied:
I've changed the license for all files to this:
```
/*
 * tools/crc32.c
 * Purpose: calculate and print POSIX checksums for a specific string.
 *
 * This tool is a clone of cksum tool from Linux coreutils.
 * Is using the same crc32 lookup table crctab as coreutils.
 * Is necessary for ifupdown-ng when the cksum tools is not avaiable on
 * a system and we need VRRP name generation.
 *
 * Copyright (c) 2023 EasynetDev in the Public Benefit, Inc.
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * This software is provided 'as is' and without any warranty, express or
 * implied.  In no event shall the authors be liable for any damages arising
 * from the use of this software.
 */

#define _GNU_SOURCE
```

I will change the the sources to get a string up to 64 bytes as an argument. In this way I will change the sources enough to be a clone of cksum. Is ok like this?

awilfox commented on tools/crc32.c:
```diff
@@ -0,0 +1,119 @@
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
```
As above.

EasyNetDev replied:
Please check my comment above and add your comments.

awilfox commented on tools/crctab.c:
```diff
@@ -0,0 +1,436 @@
+#include <stdint.h>
```
There's no license declaration at the top of this file.

Also, this code doesn't match the code style of the rest of the repository.

EasyNetDev replied:
`crctab.c`, if you check the cksum tool `cksum.c`, it is generated in case you use -DCRCTAB:

```
/* Written by Q. Frank Xia, qx@math.columbia.edu.
   Cosmetic changes and reorganization by David MacKenzie, djm@gnu.ai.mit.edu.

  Usage: cksum [file...]

  The code segment between "#ifdef CRCTAB" and "#else" is the code
  which calculates the "crctab". It is included for those who want
  verify the correctness of the "crctab". To recreate the "crctab",
  do something like the following:

      cc -I../lib -DCRCTAB -o crctab cksum.c
      crctab > crctab.c

  This software is compatible with neither the System V nor the BSD
  'sum' program.  It is supposed to conform to POSIX, except perhaps
  for foreign language support.  Any inconsistency with the standard
  (other than foreign language support) is a bug.  */
```
The code is the same over all coreutils tools. Any ideas how to solve this issue?

awilfox commented on Makefile:
```diff
@@ -87,6 +88,12 @@ YAML_SRC = \
 LIBIFUPDOWN_${CONFIG_YAML}_OBJ += ${YAML_SRC:.c=.o}
 CPPFLAGS_${CONFIG_YAML} += -DCONFIG_YAML
 
+# enable VRRP crc32 builtin tool (+31 KB). This is stripped version of cksum from coreutils
+CONFIG_CRC32 ?= Y
+CRC32_SRC = tools/crc32.c tools/cksum.c tools/crctab.c
+CRC32_${CONFIG_CRC32}_OBJ += ${CRC32_SRC:.c=.o}
+CRC32_${CONFIG_CRC32} += crc32sum
+
```
This configuration code is different from all the others.  It could be simplified, and indeed since there is no way to merge the coreutils version of `cksum` anyway the new version should be made to look like the other applets.

EasyNetDev replied:
I really don't understand how should look as other applets. Do you want to be integrated in ifupdown-ng tool and to looks like an applet like ifup / ifdown / ifquery / ifparse ?

EasyNetDev replied:
Ok. I started to port the tool as applet of ifupdown:

I changed Makefile:
```
# enable VRRP CRC32 builtin tool (+31 KB). This is stripped version of cksum from coreutils
CONFIG_CRC32 ?= Y
CRC32_SRC = cmd/crc32.c cmd/cksum.c cmd/crctab.c
MULTICALL_${CONFIG_CRC32}_OBJ += ${CRC32_SRC:.c=.o}
CMDS_${CONFIG_CRC32} += crc32sum
```

awilfox commented on Makefile:
```diff
@@ -123,21 +131,25 @@ EXECUTOR_SCRIPTS_NATIVE ?=
 LIBS += ${TARGET_LIBS} ${LIBBSD_LIBS}
 
-all: ${MULTICALL} ${CMDS}
+all: ${MULTICALL} ${CMDS} ${CRC32_Y}
```
Shouldn't be necessary.

EasyNetDev replied:
Solved. Tool is an applet now.

awilfox commented on Makefile:
```diff
@@ -123,21 +131,25 @@ EXECUTOR_SCRIPTS_NATIVE ?=
 ${MULTICALL}: ${TARGET_LIBS} ${MULTICALL_OBJ}
 	${CC} ${LDFLAGS} -o $@ ${MULTICALL_OBJ} ${LIBS}
 
+${CRC32_Y}: ${CRC32_Y_OBJ}
+	${CC} ${LDFLAGS} -o $@ ${CRC32_Y_OBJ}
+
```
As above, this should be removed and simplified.

awilfox commented on doc/interfaces-vrrp.scd:
```diff
@@ -0,0 +1,199 @@
+interfaces-vrrp(5)
```
This file starts with line breaks at a certain column, then trails off to a wider column, then appears to have no determinate column for line breaks at all.

EasyNetDev replied:
Ok, I can fix this.

EasyNetDev replied:
> This file starts with line breaks at a certain column, then trails off to a wider column, then appears to have no determinate column for line breaks at all.

Can you explain me better what do you mean with this? I was looking to my file and this file `interfaces-vxlan.scd` and I don't see any differences. My file is based on `interfaces-vxlan.scd`.

awilfox commented on doc/interfaces-vrrp.scd:
```diff
@@ -0,0 +1,199 @@
+*Mandatory*: _no_++
+*Default*: _none_
+
+# EXAMPLES using crc32sum tool and FRRouting to configure the IP interfaces
```
There should be one EXAMPLES heading, with the different examples as subheadings.

EasyNetDev replied:
Fixed in the new PR.

awilfox commented on executor-scripts/linux/vrrp:
```diff
@@ -0,0 +1,347 @@
+	crc32_tool="MOCK"
+	return
+    fi
+    crc32_cmd=$(which crc32sum)
```
`which` is not portable, use `command -v` or similar.

EasyNetDev replied:
Ok, I will fix this.

awilfox commented on executor-scripts/linux/vrrp:
```diff
@@ -0,0 +1,347 @@
+	return
+    fi
+    # Search for cksum
+    crc32_cmd=$(which cksum)
```
As above.

EasyNetDev replied:
All crc32_cmd I've changed them to :
```
    crc32_cmd=$(command -v crc32sum)
```
or
```
    crc32_cmd=$(command -v cksum)
```

awilfox commented on executor-scripts/linux/vrrp:
```diff
@@ -0,0 +1,347 @@
+
+# Compute CRC32 value using the selected CRC32 tool
+compute_crc32_hex() {
+    case "$crc32_tool" in
+	crc32sum|cksum)
+	    printf "%08x" $(echo -n "$1" | ${crc32_cmd} | cut -f1 -d" ")
+	    ;;
+	MOCK)
+	    # In test mode, check if we have crc32sum compiled, cksum installed or just return CRC32 for "eth0"
+	    if [ -x "./crc32sum" ]; then
+		printf "%08x" $(echo -n "$1" | .//crc32sum | cut -f1 -d" ")
+	    elif [ -x $(which cksum) ]; then
+		printf "%08x" $(echo -n "$1" | $(which cksum) | cut -f1 -d" ")
+	    else
+		echo "43bfa0f3"
+	    fi
+	    ;;
+	*)
+	    echo "Invalid CRC32 tool: $crc32_tool! Please use crc32sum or cksum!"
+	    exit 0
+	    ;;
```
This prevents someone from overriding the tool, for instance debugging or writing a different tool.  Perhaps even a custom tool that uses some CPU acceleration.

Should be simplified with just something akin to "if MOCK echo 43bfa0f3", "if crc32_cmd is unset error out", otherwise use the crc32_cmd as provided.

EasyNetDev replied:
Ok, could be a solution. I can change this part of the script.

EasyNetDev replied:
Removed the additional code and changed to this:
```
# Compute CRC32 value using the selected CRC32 tool
compute_crc32_hex() {
    case "$crc32_tool" in
        crc32sum|cksum)
            printf "%08x" $(echo -n "$1" | ${crc32_cmd} | cut -f1 -d" ")
            ;;
        MOCK)
            # This is CRC32 hex value calculated for "eth0" string.
            echo "43bfa0f3"
            ;;
        *)
            echo "Invalid CRC32 tool: $crc32_tool! Only allowed tools are crc32sum or cksum!"
            exit 0
            ;;
    esac
}
```

awilfox reviewed (changes requested):
Thank you for your contribution!  At this time, this PR is not mergeable in its present state.  Please review the inline comments provided.

EasyNetDev commented:
> @EasyNetDev: Do not forget to update the PR to have a quick review from @ifupdown-ng!

Yes. As soon I'm finishing crc32 code as an applet I will push the new code.

EasyNetDev commented:
Let's recap the changes:

1. Moved the code to `cmd`
2. Built it as applet
3. Add `-X | --hex` option to directly output hex value of the CRC32 for VRRP executor.
4. The new command is called ifcrc32sum:
```
$ ./ifcrc32sum
ifcrc32sum - calculate CRC32 value for VRRP MACVLAN subinterfaces.

This tool is a backup tool in case cksum from coreutils is not available.
The argument <interface> must have up to 64 characters. Everything over 64 charcters are ignored in calculation.

Usage:
  ifcrc32sum <interface>

Global options:
  -h, --help                          this help
  -V, --version                       show this program's version

Program-specific options:
  -X, --hex                           print CRC32 value in hex for the interface

For more information: man 8 ifcrc32sum
```
6. Added doc/ifcrc32sum.scd

To do:
1. Change the VRRP code to use the new tool
2. Change Makefile to build document.

Any other suggestions?
@awilfox could you provide me additional info about my `doc/interfaces-vrrp.scd` ? All `scd` files are following the same rules.

Neustradamus commented:
@EasyNetDev: Why have you removed it?

EasyNetDev commented:
This is the new version with applet: https://github.com/EasyNetDev/ifupdown-ng/tree/ifupdown-ng/vrrp-1.7

Neustradamus commented:
@EasyNetDev: You have not done a new PR

EasyNetDev commented:
> @EasyNetDev: You have not done a new PR

Not yet. I'm pushing to my main Debian branch, compile and test it on a machine with multiple VRRP interfaces and check if I find some bugs before push the PR for v1.7

EasyNetDev commented:
New PR pushed as #237.

Neustradamus commented:
@EasyNetDev: You can close now ;)

EasyNetDev commented:
Closed via #237
